File: partiql-spi/src/main/java/org/partiql/spi/function/Accumulator.java
package org.partiql.spi.function;

import org.jetbrains.annotations.NotNull;
import org.partiql.spi.value.Datum;

/**
 * Represents the accumulator for an aggregation function.
 * @see AggProvider
 * @see Agg
 */
public interface Accumulator {

    /**
     * Receives the next set of arguments, and updates the accumulator's state.
     * @param args the next set of arguments.
     */
    public void next(Datum[] args);

    /**
     * Computes the final value of the accumulator.
     * @return the computed final value of the accumulator.
     */
    @NotNull
    public Datum value();
}

File: partiql-eval/src/main/kotlin/org/partiql/eval/internal/Accumulator.kt
package org.partiql.eval.internal

import org.partiql.eval.value.Datum

/**
 * Aggregation function state.
 *
 * TODO consider a `done()` method for short-circuiting.
 */
interface Accumulator {

    /**
     * Apply args to the accumulator.
     *
     * @param args
     * @return
     */
    fun next(args: Array<Datum>)

    /**
     * Return the accumulator value.
     *
     * @return
     */
    fun value(): Datum
}
